src/System.Text.Json/src/System/Text/Json/Serialization/JsonSnakeCaseNamingPolicy.cs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
+    {
+        public override string ConvertName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+                return name;
+
+            // Allocates a string builder with the guessed result length,
+            // where 5 is the average word length in English, and
+            // max(2, length / 5) is the number of underscores.
+            StringBuilder builder = new StringBuilder(name.Length + Math.Max(2, name.Length / 5));
+            UnicodeCategory? previousCategory = null;
+
+            for (int currentIndex = 0; currentIndex < name.Length; currentIndex++)
+            {
+                char currentChar = name[currentIndex];
+                if (currentChar == '_')
+                {
+                    builder.Append('_');
+                    previousCategory = null;
+                    continue;
+                }
+
+                UnicodeCategory currentCategory = char.GetUnicodeCategory(currentChar);
+
+                switch (currentCategory)
+                {
+                    case UnicodeCategory.UppercaseLetter:
+                    case UnicodeCategory.TitlecaseLetter:
```

Comment:
I don't see why we need to special case TitlecaseLetter:

https://www.compart.com/en/unicode/category/Lt

Reply:
Do you mean that they cannot be lowercased? I haven't worked with Unicode a lot, so it's out of my knowledge.

Reply:
I am also not super familiar with all the categories. I am asking if you had a test case or scenario that motivated including `TitlecaseLetter` in the implementation to begin with. It looks like they do have lower-case counterparts. Again, adding tests where we exercise this `UnicodeCategory` more would be useful. I believe, if we remove this switch case, all the existing tests would still pass, which means there is a test gap.

@GrabYourPitchforks - do you have some thoughts on the UnicodeCategories we are special casing here for snake casing?

Reply:
I am primarily asking for adding more tests so that we can more confidently optimize the implementation later. I want to avoid removing some code path in the future to improve perf that would end up breaking functionality.

Reply:
Checked. They all can be lowered.